src/test/java/com/google/devtools/build/lib/remote/downloader/GrpcRemoteDownloaderTest.java

```diff
@@ -0,0 +1,314 @@
+    final byte[] downloaded = downloadBlob(
+        downloader, new URL("http://example.com/content.txt"), Optional.<Checksum>empty());
+
+    assertThat(downloaded).isEqualTo(content);
+  }
+
+  @Test
+  public void testPropagateChecksum() throws Exception {
+    final byte[] content = "example content".getBytes(UTF_8);
+    final Digest contentDigest = DIGEST_UTIL.compute(content);
+
+    serviceRegistry.addService(
+        new FetchImplBase() {
+          @Override
+          public void fetchBlob(
+              FetchBlobRequest request,
+              StreamObserver<FetchBlobResponse> responseObserver) {
+            assertThat(request).isEqualTo(FetchBlobRequest.newBuilder()
+                .addUris("http://example.com/content.txt")
+                .addQualifiers(Qualifier.newBuilder()
+                    .setName("checksum.sri")
+                    .setValue("sha256-ot7ke6YmiSXal3UKt0K69n8C4vtUziPUmftmpbAiKQM="))
+                .build());
+            responseObserver.onNext(
+                FetchBlobResponse.newBuilder().setBlobDigest(contentDigest).build());
+            responseObserver.onCompleted();
+          }
+        });
+
+    final RemoteCacheClient cacheClient = new InMemoryCacheClient();
+    final GrpcRemoteDownloader downloader = newDownloader(cacheClient);
+
+    getFromFuture(cacheClient.uploadBlob(contentDigest, ByteString.copyFrom(content)));
+    final byte[] downloaded = downloadBlob(
+        downloader, new URL("http://example.com/content.txt"),Optional.<Checksum>of(
+            Checksum.fromString(KeyType.SHA256, contentDigest.getHash())
+      ));
+
+    assertThat(downloaded).isEqualTo(content);
+  }
+
+  @Test
+  public void testRejectChecksumMismatch() throws Exception {
+    final byte[] content = "example content".getBytes(UTF_8);
+    final Digest contentDigest = DIGEST_UTIL.compute(content);
+
+    serviceRegistry.addService(
+        new FetchImplBase() {
+          @Override
+          public void fetchBlob(
+              FetchBlobRequest request,
+              StreamObserver<FetchBlobResponse> responseObserver) {
+            assertThat(request).isEqualTo(FetchBlobRequest.newBuilder()
+                .addUris("http://example.com/content.txt")
+                .addQualifiers(Qualifier.newBuilder()
+                    .setName("checksum.sri")
+                    .setValue("sha256-ot7ke6YmiSXal3UKt0K69n8C4vtUziPUmftmpbAiKQM="))
+                .build());
+            responseObserver.onNext(
+                FetchBlobResponse.newBuilder().setBlobDigest(contentDigest).build());
+            responseObserver.onCompleted();
+          }
+        });
+
+    final RemoteCacheClient cacheClient = new InMemoryCacheClient();
+    final GrpcRemoteDownloader downloader = newDownloader(cacheClient);
+
+    getFromFuture(cacheClient.uploadBlob(contentDigest, ByteString.copyFromUtf8("wrong content")));
+
+    IOException e = assertThrows(UnrecoverableHttpException.class, () -> downloadBlob(
```

Comment:
Fix formatting to improve readability:

    IOException e = assertThrows(
        UnrecoverableHttpException.class,
        () -> downloadBlob(
            downloader,
             ...,
             ...));

Reply:
Done in forked PR https://github.com/bazelbuild/bazel/pull/10914